### DISCUSSION TITLE
[4.1] Ad-hoc single asset indexing

### DISCUSSION THREAD

**Discussion by rob-c-baker** [craftcms/cms#11998] [Q&A] [answered]
I have a piece of custom module code _(from Craft 3 that I am looking to update for Craft 4)_ that produces an image thumbnail from a PDF, sets the resulting thumbnail as an asset in a specific volume / folder and then asks the `AssetIndexer` to index the image. The "parent" PDF would often change so when this code is invoked it needs to get the asset index updated (i.e. load an asset, swap out it's path / filename, reindex)

In the Craft 3 days, this was done like this:

```php
$vol = $asset->getVolume();
$path = $asset->getFolder()->path . $new_filename;
$indexed_asset = Craft::$app->getAssetIndexer()->indexFile($vol, $path);
```

But since Craft 4, a `$sessionId` parameter is needed like this as I presume it is assumed this will be run from a CLI context:

```php
$indexed_asset = Craft::$app->getAssetIndexer()->indexFile($vol, $path, $sessionId);
```

I can roughly see how a `$sessionId` is arrived at in the `AssetIndexer` service but it's complicated.

I would love an ad-hoc way of (re)indexing a single asset without the need for the `$sessionId` and I cannot find a good way to do that in v4.

Does anyone have any ideas how to do this, or can we please have a new method on the `AssetIndexer` for this purpose?

**Comment by brandonkelly** [comment, marked as answer]
I’d recommend you just stick to the normal `assets` service API. Any changes made with that will automatically be indexed.

```php
use Craft;
use craft\elements\Asset;
use craft\helpers\Assets;

$tempPath = '/path/to/new/thumb.png';
$volumeHandle = 'volumeHandle';
$folderPath = 'the/folder/path';

$filename = Assets::prepareAssetName($tempPath);
$volume = Craft::$app->volumes->getVolumeByHandle($volumeHandle);
$folder = Craft::$app->assets->ensureFolderByFullPathAndVolume($folderPath, $volume);

$asset = Asset::find()
    ->volume($volume)
    ->folderId($folder->id)
    ->one();

if ($asset) {
    // Replace the existing thumb with the new one
    Craft::$app->assets->replaceAssetFile($asset, $tempPath, $filename);
} else {
    // Save as a new asset
    $asset = new Asset();
    $asset->tempFilePath = $tempPath;
    $asset->setFilename($filename);
    $asset->newFolderId = $folder->id;
    $asset->setVolumeId($volume->id);
    $asset->setScenario(Asset::SCENARIO_CREATE);

    if (!Craft::$app->elements->saveElement($asset)) {
        // didn't work. check $asset->getErrors().
    }
```

**Reply by rob-c-baker** [reply]
Thanks, I can confirm this works great.

I'm _manually_ creating the thumb image from a PDF and then putting the file into a path returned by `Craft::$app->getAssets()->getUserTemporaryUploadFolder()` - that then goes into the `$tempPath` parameter above and everything works!

There was a reason I was using the `AssetIndexer` first time around on v3 but I can't recall why and the commit messages don't help!
